### DISCUSSION TITLE
Propagation of CultureInfo to a nested interpolated string

### DISCUSSION THREAD

**Discussion by ufcpp** [dotnet/csharplang#6882] [Language Ideas]
## TL;DR

I'd like a mechanism to propagate `CultureInfo` to a nested interpolated string with the following code.

```cs
Thread.CurrentThread.CurrentCulture = CultureInfo.InvariantCulture;

string.Create(
    CultureInfo.GetCultureInfo("fr-fr"), 
    $"direct: {1.5}, {$"nested: {1.5}"}"));
```

## Background

I am writing [a utility library for string interpolation](https://github.com/ufcpp/StringInterpolation). One of the features provided by this library is the ability to do the following

```cs
static string format(bool condition, params int[] days) => $$"""
{
  "a": {{If(condition,
    b => b.Append($"{123}"),
    b => b.Append($"{1.5}")
    )}},
  "b": [{{ForEach(days, (v, b) => b.Append($"""

    "{new DateOnly(2000, 12, v)}",
"""))}}
  ]
}

""";
```

However, this code has the following issue.

### IFormatProvider propagation

Unfortunately, string manipulation in .NET mostly depends on `CurrentCulture`. There are many situations where this is not desirable, and the culture is required to be explicitly specified.

For string interpolation, you would use `string.Create(culture, $"")` instead of `$""`. However, the `culture` explicitly passed here is not propagated to the nested interpolated string.

For example, execute the following code:

```cs
using System.Globalization;

Thread.CurrentThread.CurrentCulture = CultureInfo.InvariantCulture;

Console.WriteLine(
    string.Create(
        CultureInfo.GetCultureInfo("fr-fr"), 
        $"""
        direct: {1.5}
        {$"nested: {1.5}"}
        """));
```

The following results are obtained:

```cs
direct: 1,5
nested: 1.5
```
It would be a solution if `string.Create(culture)` were made explicit again in the `nested` part, but human error inevitably occurs.

Returning to [my library](https://github.com/ufcpp/StringInterpolation), `If` and `ForEach` in this library propagate `culture` over the argument of the lambda expression `b`. As a result, the boilerplate code `b => b.Append($"")` is required. This is cumbersome, and I would prefer to write it without `b.Append`, as shown below.

```cs
static string format(bool condition, params int[] days) => $$"""
{
  "a": {{If(condition, $"{123}", $"{1.5}")}},
  "b": [{{ForEach(days, v => $"""

    "{new DateOnly(2000, 12, v)}",
""")}}
  ]
}

""";
```

However, in the current version of C#, I can't think of a way to propagate `culture` in this style.

A close implementation, as far as it can be written in C# 11, would be as follows.

* [Gist](https://gist.github.com/ufcpp/9d885ec3f0e62d78dea2b248a272d0cb)
* [sharplab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQHEA2A9gEYCG+AlgF6kAuFhAdgNzJqY4AqAFgE4CmpYBUYBzVigxZsAJQCujegFt+2AMKElABwr5+vAMr6AbhQDG/AM4TkPAUPVzeAxXcHBHz/orVz8tJ350AF50X39AgElGADNCbGjjUl4KUh8/AIEbFDgANg4MJQAKYkJCfHQzJmF6JgBKEIA+AvV7Wn4i5HRusIyo2PjcflpwzP5ouKKAIhjeAFpZqbqAGnQunoASKeEBM1oQdABvHABWAF90Ql4juBAzM9XGK3bQI8iYoqrGGoZGVa3jthzlN/lNjiA1GcllC6tksABOIrFWi8OT8OqwthwRHFGLkSzozFILH5LAYd4lMoVL4/JjLdbdADa0XavC05To/GABhRIlEAAk0sA9LwAIK8URyFSKIqMUgqQgfGkUWqMDEAXXQ704qPGin07PwnO5vLEgu+IvQtG43nNwv09KQPXQzP1bI5Lx5KTNQpF4sl0tosvl/EVn2qKt+Gq1MQAYvi9azDcavXy7Zb+PgCen9HUGU1KhHVegAPxWm2MHO8bCcQipsSi75qPTJIoNA6Z7O+/Q1uum0SN4DNwS8NvZV1Jj1c+sC7u8dVsADM6AEMXQlhRcj2MZ1aJZBqnJu9s4t+mQh3zWhSSXa6AAIvw8Rl9+6jZ7+1X0AB9ERPXjZZ0oGXbVdRfZN32PKsihEWh0EoVlyAAGW8URrVWGD0DiXglDoF4NAUWhVlKcpC2+SM6UuORYOIipLG4Qg/GAUUtC0bxgDzJ0egvTjnW6OiGPwJiWLYkJSNpFh82dCg1yKABCZVVQaKAAHYJF4nof0YP9RKeAB3e9H1IZ83XA6cPznIp4P0JCULQzCrhw2g8IYxQiWdM5kHzICd1AkzDxnKCMKs3gbLEOyMKwxznIIoiqTE8i/ko6i4v4xjmNY75VkiWMHLoAAFXhCFMYB9DLK8iooEreA451uPUvj6LS4TvlEhTfjU+rpPQOS2vqTBVMkjTf30HT+H0h8n38MD/PM09R2C0LUO4VZItwrl8MUVZyuK3MOu6DziR4zBlzJdBaxnQdh1bBpgmaTS/17C6mxbUciS8k6ABZ0HStjEJVaz8CKU6knwNEbru4bqx+74/oQwGQbBvbjswL7oeAHLsLW4AgbgVAywR9EC3uns0YxqKuSKAm3J6byoFR5r0dypyKcsKpWOAdAZHcAB5Rh8AATwMLQ0gAHjMbhkmaKmich7BSaZl5KfIRH3pR76GbJrGihIAArfg9nx5X+HQxR0HIChREYUTUFWMky1W2DQkYPx8HB79ZflzHmexgnVnNy2VqZ6nulp+mMsZr3Fbt9BffQDD/at0IbYKe2mZ0l23eJqGNYVinY4TwOveD5G6fV8PNe9opWcIdnOZ5vnBeFxgxYl3gpaNk3YIT63bdx1OvfT/BXZlrSSZzyO847s3KAD+yi6R0Oy7YiuXhFzhGiKTgY6NzOPfH8mfZ3hfPqX74V65NeN632Po4d3fR+z8vc8P0Hjbnxzi8Xz2D8vzft9fzu08LZqhHg9b+Wt84zzVMfNW4Dva/2vlPeOUDe543fnQe+YD94QKngXdBtAiQHWQBOA8b4zKQTnAuFAy5Vzrk3Nud48YsyJlISmWa9peDnkvNeTkBlJq0GmmQo8aY5zuwfgBGmwE4wJkEWwihc1oKmwWvgZCYVlpx1Ng7aKm10A0XiqqVYDFkokVSoJNGNUuKDQagJIS4dRLySLO1KxccZK9RASpJGzos6jXGoZYyk4hEBQsso1RS1C4Hw2gQpGRCjreUYTIvygT2EikUbBEJtl1ERWfpE2KJE3GGKorolKjUzEMyyhXAqFUqplUKjtaq+Y6r1VMbYkSoQHFkVVJ4noXVwwdKjP1Lp3RvFOzGnwoyU1ElyJEQo9JajwlYxyegbalVdr5hiYBE6uMzp9mPJdF6bZQE9nOv2PZI4xyeViSfNGsMAY4wwNLW6YisHhxuSFeGR9ValzgVHPu/8waHMfsvZ+StX6fyudgyu1da5cyELzAWQtRbi0ln8wmjys5ywhYrKmMCvmYoprrfWtBDYAI0V3KBPcU74MHsPNFe8n4TxfmiP2KD8FgtgXi7G0dY7IOARS2+acnYZwBRi+lB8QVMqAbPO+OKw5AoZVXNmXI66wobgi5uSK24osAd3JOqD+6OWpZgseoqcEkrwdKz5sqz7PwQSio1gLrXyuxZa0+Ecf7rz/jfX5d9hXfMniSi1lz2UmvgR6xBJKeWW3tSKuVYrIHALZbikNq8w1atJZKxK/Ki6+o5eKt+5qg4SAOkAA)

Here is an excerpt:

```cs
Thread.CurrentThread.CurrentCulture = CultureInfo.InvariantCulture;

static string m(bool condition) => string.Create(
    CultureInfo.GetCultureInfo("fr-fr"), 
    $"direct: {1.5} or {1:C}, nested: {If(condition, $"{1.5}", $"{1:C}")}");

Console.WriteLine(m(true));
Console.WriteLine(m(false));
```

The current implementation produces the following output. (Different for direct and nested.)

```
direct: 1,5 or 1,00 €, nested: 1.5
direct: 1,5 or 1,00 €, nested: ¤1.00
```

The expected output is rather like this
(Same with direct and nested.)

```
direct: 1,5 or 1,00 €, nested: 1,5
direct: 1,5 or 1,00 €, nested: 1,00 €
```

I am having trouble finding a way to do that in the current C#.

The problem is that there is no way to propagate the `CultureInfo` passed to `string.Create` to the handler of the argument of the `If`.

## Discussion

Now, the aforementioned `string.Create` code lowers to the following code

```cs
IFormatProvider cultureInfo = CultureInfo.GetCultureInfo("fr-fr");
IFormatProvider provider = cultureInfo;
DefaultInterpolatedStringHandler handler = new(22, 3, cultureInfo);
// ...
IfTrueInterpolatedStringHandler thenHandler = new(0, 1, condition, out var shouldAppend);
if (shouldAppend)
{
    thenHandler.AppendFormatted(1.5);
}
IfFalseInterpolatedStringHandler elseHandler = new(0, 1, condition, out shouldAppend);
if (shouldAppend)
{
    elseHandler.AppendFormatted(1, "C");
}
handler.AppendFormatted(If(condition, thenHandler, elseHandler));
return string.Create(provider, ref handler);
```

Can we somehow propagate `cultureInfo` to the nested handlers as follows?

```cs
IFormatProvider cultureInfo = CultureInfo.GetCultureInfo("fr-fr");
IFormatProvider provider = cultureInfo;
DefaultInterpolatedStringHandler handler = new(22, 3, cultureInfo);
// ...
IfTrueInterpolatedStringHandler thenHandler = new(0, 1, condition, out var shouldAppend,
    cultureInfo // !!!
    );
if (shouldAppend)
{
    thenHandler.AppendFormatted(1.5);
}
IfFalseInterpolatedStringHandler elseHandler = new(0, 1, condition, out shouldAppend,
    cultureInfo // !!!
    );
if (shouldAppend)
{
    elseHandler.AppendFormatted(1, "C");
}
handler.AppendFormatted(If(condition, thenHandler, elseHandler));
return string.Create(provider, ref handler);
```

To achieve this, could we use a mechanism such as [Implicit parameters](https://github.com/dotnet/csharplang/issues/6300)?

```cs
[InterpolatedStringHandler]
public ref struct IfTrueInterpolatedStringHandler
{
    public IfTrueInterpolatedStringHandler(int literalLength, int formattedCount, bool condition, out bool shouldAppend,
        implicit IFormatProvider? provider) // or some attributes like InterpolatedStringHandlerArgument
    {
        shouldAppend = condition;
        if (!condition) return;
        _inner = new DefaultInterpolatedStringHandler(literalLength, formattedCount, provider);
    }
```

"[implicit hell](http://spiridonov.pro/2015/10/14/scala-implicit-hell/)" is scary, but IMHO, implicit parameters are somewhat better than relying on static properties (like `CurrentCulture` or `SynchronizationContext.Current`) 

I would like to hear your opinions.

**Comment by 333fred** [comment]
This is kinda already possible, by defining a `AppendFormatted` method that takes the containing handler type and copying over any relevant data from the original handler. Here's an example: https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQiCM65IARyYB0A4gDYD2ARgIZkCWAXjcPRVANw76FEBK0VgFsApkQDCFIQAd6ZEWADKCgG70AxiIDOXDD2IAVEQA9gunJgCcACnEQtwKQAkaUACbywEsCJYjbEGTAED4AklAAZhSkIsB2QSEi4VHWAEQRYAC0GakAlAA0eOh4JXgAJKlu9D7qoHgA3sQArAC+hfUVUNrAIm4gDc0tqUO5ueZoANrhPWDSFGR+borAYPRQAOYu7p4AuugOkLV4dg7Orh4K6PXFpXAAzASYAGyP+OI+fgEJYZEUeGtRQpTKAzOYLHpLFZrTbnTwAQTA6wgohB1igNFEFAi1gBFFGO2O9kcQi2FzAeHURLO2wUuRuJWuaFKzIIAHYKVSSbCFEQqBA5G4eQYKMtVhtrLkAIS6ZktelFJmlaSrFR+PAfNzsMgATzwouhACF+WRBeS+QKFHgALx4LoAdwlMqVKrVGq1uviwR+UQ53xETpK8vuhNOXJpYBxILwDBmdAAMiINsAABaFNbAPBRMBCFgQySCQqexLJP6Uv0lumKhny5llr0ia2++slgN4OVVhXM4MnYmkzyRjMxhTxxPrFNpqNZnPAPMUAsh3vc8nJpe5PD9FP0LTWIdgEdJ1OZijZ3O9fMgwor8NEOuJSvMxmy+XPjvBuAAFjwcOk0kTbjj9CxmQ1iEPgWj3qUj4siU5omjy36/u41jga27bMkGDwfl+P5/gAYse04QtYwKgvMiz6hsfYKAiSIosAaR5ASPbUmSeBXmSEHVh2zJZr46jJng1iquS/HQAA1v8UBsUuvLGqaMRxMm4nbqMNaQWp0GwfJCF/tYolQGJYwaW28poaUGEEJ+OnuPhJ4zr01iahAVDyHgbicQ0xlafBOFIW4RDChR6x6YE9ajKh6AtEAA=.

Unfortunately, I don't think that's immediately useful for your library, as the nested strings _aren't_ direct arguments of containing string, but are instead arguments of a nested function of the string. If you were willing to incur some boxing and allocations, you could take advantage of this trick by having a `NestedBuilder` type that stores the original values, and then calls `AppendFormatted` for each value when appended to the containing handler. Somewhat complicated and expensive, to be sure, but it would work for the scenario.

**Reply by ufcpp** [reply]
It may take some time to organize the ideas, so please allow me to answer in several parts.

> If you were willing to incur some boxing and allocations

No. One of the goals of my library is to avoid allocations.

I have an implementation that `Append` directly to `destination` in `TryFormat` via `ISpanFormattable` once, e.g. [`Join`](https://github.com/ufcpp/StringInterpolation/blob/main/src/StringInterpolation/Format.JoinT.cs).

**Comment by ufcpp** [comment]
Wouldn't that the `AppendFormatted` that takes the handler be insufficient for an `If `or `ForEach ` method?

```cs
CustomHandler.Create(
    CultureInfo.GetCultureInfo("fr-fr"),
    $"direct: {1.5} or {1:C}, nested: {
        If(condition, // cannot invoke AppendFormatted(CustomHandler handler) directy
        $"{1.5}",
        $"{1:C}")
        }");
```

**Reply by ufcpp** [reply]
If the string interpolation were target-typed, the problem might be alleviated a bit.

```cs
CustomHandler.Create(
    CultureInfo.GetCultureInfo("fr-fr"),
    $"direct: {1.5} or {1:C}, nested: {(condition
        ? $"{1.5}" // Since this is not target-typed, so $"{1.5}" is string instead of CustomHandler.
        : $"{1:C}")
        }");
```

**Reply by 333fred** [reply]
> Wouldn't that the AppendFormatted that takes the handler be insufficient for an If or ForEach method?

That's what I said about it requiring allocations.

> If the string interpolation were target-typed, the problem might be alleviated a bit.

It is. What you're seeing is that ternaries with a natural type (string) are not.
